### PR TITLE
Adopt custom Java style guide

### PR DESCRIPTION
**Pull request by nwithan8**
This branch is completed and ready for review, defined as:
- A custom EasyPost Java style guide has been created.
  - This is largely based on the Sun style guide, with a handful of custom modifications, including increasing the line width from 80 to 120 characters, and custom variable name regex rules that allow for _ (snake case names) (required for smartrate percentile field names).
  - The guide file is available [here](https://github.com/EasyPost/easypost-java/blob/style/easypost_java_style.xml). This is a [CheckStyle](https://checkstyle.org/) XML config file, which can be imported and used by a number of [IDE plugins](https://checkstyle.org/#Related_Tools_Active_Tools) for checking and enforcing [these style rules](https://checkstyle.org/checks.html).
  - This style guide will be a living, breathing document. I do not nor should not have the power to unilaterally decide what our EasyPost style is or should be. EasyPost developers, both internal and community, should feel free to weigh in with suggested edits to the style. Enforcing these rules (via the CheckStyle plugins) is easy; deciding what these rules should be, that's the difficult part.
- All code in this repository has been adjusted to comply with the style guide.
  - Not all rules are being enforced. Any rules in the guide can be overridden in-line (f2681b6), and entire files/directories can be excluded from checks (c98a21f)
  - Again, some minor semantic rules will need to be fleshed out by committee. For now, though, the most notable aspects of the guide are being adhered to.
 
This branch also contains several unrelated commits due to a rebase with `master` after a separate branch was merged into `master`. These commits will be squashed before this branch is merged into `master`.

### CODE REVIEW

**Comment by Justintime50** [comment]
Conceptually this looks really good - it cleans everything up really nicely and documents it. I'm not a huge fan of the tiny `80` character line limit, but it is what it is. We may consider overwriting that to say `120` which is our standard internally and for other languages because it's much more manageable.

I think with this one it will be super important to ensure we get the `private/public` definitions correct as that will start limiting/opening their availability like they haven't been to this point. 

You have the green light from me to continue on this path - looked through the PR and it looks good. We can fix any small items during the real code review once it's finished but I didn't find any glaring things to tweak yet.

**Comment by fiendish** [comment]
Please ditch the stone-age 80 character line wrap. Google's style guide would be a bit more modern.

**Comment by nwithan8** [comment]
Created custom EasyPost Java CheckStyle configuration (largely based on Sun): https://github.com/EasyPost/easypost-java/blob/76224f5281f1482d2c36361ac0d142f93d78d97b/easypost_java_style.xml

Documentation: https://checkstyle.org/checks.html

This XML file can be imported into a CheckStyle IDE plugin for real-time checks and validation: https://checkstyle.org/index.html#Related_Tools_Active_Tools

**Comment by fiendish** [comment]
> We can also use a GitHub Actions linter to enforce the style: https://github.com/marketplace/actions/run-java-checkstyle

maven build is already being run. should be able to add it to the pom file.

File: easypost_java_style.xml
Comment (fiendish) [inline easypost_java_style.xml]: Why remove the doctype declaration?

Comment (nwithan8) [reply easypost_java_style.xml]: My mistake, my IDE was complaining so I removed it. Figure out how to resolve the issue and it's been readded.

**Comment by nwithan8** [comment]
This branch is completed and ready for review, defined as:
- A custom EasyPost Java style guide has been created
  - This is largely based on the Sun style guide, with a handful of custom modifications, including increasing the line width from 80 to 120 characters, and custom variable name regex rules that allow for _ (snake case names)
  - The guide file is available [here](https://github.com/EasyPost/easypost-java/blob/style/easypost_java_style.xml). This is a [CheckStyle](https://checkstyle.org/) XML config file, which can be imported and used by a number of [IDE plugins](https://checkstyle.org/#Related_Tools_Active_Tools) for checking and enforcing [these style rules](https://checkstyle.org/checks.html)
  - This style guide will be a living, breathing document. I do not nor should not have the power to unilaterally decide what our EasyPost style is or should be. EasyPost developers, both internal and community, should feel free to weigh in with suggested edits to the style. Enforcing these rules (via the CheckStyle plugins) is easy; deciding what these rules should be, that's the difficult part.
- All code in this repository has been adjusted to comply with the style guide.
  - Not all rules are being enforced. If you were to run a CheckStyle check on the repository, it would flag several lines, almost all of which deal with fields being public rather than private. This is a known "issue" as a result of how we handle JSON de/serialization, which requires many of our fields to be public rather than private. For now, this is fine. In the future, we should alter either the style guide or the code itself when a decision has been made about this.
 
This branch also contains several unrelated commits due to a rebase with `master` after a separate branch was merged into `master`. These commits will be squashed before being merged into `master`.

**Review by Justintime50 — DISMISSED** [review]
First off, incredible work here @nwithan8, this was no easy task!

Most of this looks great! I found a couple things we can help clarify for those who come after us. I had a couple of questions - but overall the work done here is fantastic.

File: easypost_java_style.xml
Comment (nwithan8) [reply easypost_java_style.xml:96]: Was that way in the original Sun config. Vote to remove it?

Comment (Justintime50) [reply easypost_java_style.xml:96]: Yep, I see no use in keeping blocks of commented out code unless we immediately planned to do something with it in the next PR. Kill it.

File: src/main/java/com/easypost/model/PostageLabel.java
Comment (nwithan8) [reply src/main/java/com/easypost/model/PostageLabel.java:342]: Oops, GitHub Co-Pilot got ahead of itself.

File: src/main/java/com/easypost/model/Rate.java
Comment (nwithan8) [reply src/main/java/com/easypost/model/Rate.java:358]: CheckStyles hates this constructor because we're passing in more than 7 parameters.
I removed it and reworked our RateDeserializer to create it more cleanly.

File: src/main/java/com/easypost/model/SmartrateCollection.java
Comment (nwithan8) [reply src/main/java/com/easypost/model/SmartrateCollection.java:19]: It's there by mistake.
I know, because I added it.

File: src/main/java/com/easypost/net/EasyPostResource.java
Comment (nwithan8) [reply src/main/java/com/easypost/net/EasyPostResource.java:34]: Sorry, the rest of the work got lost in the diff.
This provides enums for HTTP status codes, so instead of comparing a code to, e.g., `404`, we compare it to `HttpUrlConnection.NOT_FOUND`
Example: https://github.com/EasyPost/easypost-java/pull/80/files#diff-562d3c98a6e71dc50980fe8300ab00742293eb925eeb5f106f0086704030519aR392

File: src/test/java/com/easypost/EventTest.java
Comment (nwithan8) [reply src/test/java/com/easypost/EventTest.java]: I cleaned it up significantly in 69e726e

File: src/main/java/com/easypost/model/TimeInTransit.java
Comment (fiendish) [reply src/main/java/com/easypost/model/TimeInTransit.java]: >     (We just Gson, which expects a one-to-one match between JSON keys and Java field names.)

Nit: one-to-one doesn't mean identical. 

More importantly, neither one-to-one nor identical matching are real expectations for gson, only the default behavior. 

See e.g. `@SerializedName`
https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/annotations/SerializedName.html
and/or
`FieldNamingPolicy` https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/FieldNamingPolicy.html

File: src/main/java/com/easypost/model/TimeInTransit.java
Comment (fiendish) [inline src/main/java/com/easypost/model/TimeInTransit.java:17]: >     (We just Gson, which expects a one-to-one match between JSON keys and Java field names.)

Nit: one-to-one doesn't mean identical. 

More importantly, neither one-to-one nor identical matching are real expectations for gson, only the default behavior. 

See e.g. `@SerializedName`
https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/annotations/SerializedName.html
and/or
`FieldNamingPolicy` `LOWER_CASE_WITH_UNDERSCORES` https://www.javadoc.io/doc/com.google.code.gson/gson/2.6.2/com/google/gson/FieldNamingPolicy.html

**Review by fiendish — CHANGES_REQUESTED** [review]
^

File: src/main/java/com/easypost/model/TimeInTransit.java
Comment (nwithan8) [reply src/main/java/com/easypost/model/TimeInTransit.java:17]: Oh, that's a good point. We could name the variables according to the style guide, and then just pop a @SerializedName annotation on them to clarify their JSON key name.

In fact, I think I'd like to do that for all fields.

Comment (nwithan8) [reply src/main/java/com/easypost/model/TimeInTransit.java]: I went ahead and added the annotations in TimeInTransit.java to allow the percentile fields to be de/serialized properly, while also allowing the field names to comply with our style guide (1152995). The getter and setter functions have been renamed, so this is a BREAKING CHANGE.

TimeInTransit was the only place where this was an issue, since all our other JSON fields either are not snake case or are handled by the FieldNamingPolicy we employ in Gson.

That being said, I think it would be smart in the future to explicitly declare the JSON key names by adding SerializedName to all fields. This would help in the future if a JSON key were to change, as we wouldn't have to change the Java field name, just update the annotation.

File: src/main/java/com/easypost/model/TimeInTransit.java
Comment (fiendish) [inline src/main/java/com/easypost/model/TimeInTransit.java:154]: I see stuff like this and I wonder why we can't just use a struct-like class with public members and no accessors.

File: src/main/java/com/easypost/model/TimeInTransit.java
Comment (fiendish) [reply src/main/java/com/easypost/model/TimeInTransit.java]: 👍  Withdrawing my block.

File: easypost_java_style.xml
Comment (fiendish) [inline easypost_java_style.xml:140]: Can this be removed now?

Comment (nwithan8) [reply easypost_java_style.xml]: Yes, it has been in fe93df6

**Review by fiendish — APPROVED** [review]

File: easypost_java_style.xml
Comment (Justintime50) [reply easypost_java_style.xml]: I suppose the link to documentation is enough here.

File: easypost_java_style.xml
Comment (Justintime50) [reply easypost_java_style.xml]: I suppose the link to documentation is enough here.

File: easypost_java_style.xml
Comment (Justintime50) [reply easypost_java_style.xml]: Thanks, good comment.

File: src/main/java/com/easypost/model/Fee.java
Comment (Justintime50) [inline src/main/java/com/easypost/model/Fee.java:53]: These are similarly confusing as the `refund` verbiage was below. Can we tweak these too so the description doesn't sound conditional for the getter and setter?

File: src/main/java/com/easypost/model/SmartrateCollection.java
Comment (Justintime50) [reply src/main/java/com/easypost/model/SmartrateCollection.java:20]: It looks like this went from an empty Collection to containing logic? Why is that? If it needed logic, can we call this change out in the summary of the PR? If this was an accident let's fix it so it stays the same.

If it now contains logic, there's no reason now to keep the `intentional` comment as it's no longer empty.

File: src/main/java/com/easypost/net/EasyPostResource.java
Comment (Justintime50) [reply src/main/java/com/easypost/net/EasyPostResource.java:34]: Ah I see, thanks for clarifying.

File: src/test/java/com/easypost/EventTest.java
Comment (Justintime50) [reply src/test/java/com/easypost/EventTest.java]: Dang... looks nice but is a decently large change. If tests still pass I'll just move on ha.

**Review by Justintime50 — COMMENTED** [review]
Looks great! Left two questions, then I'll approve.

File: src/main/java/com/easypost/model/SmartrateCollection.java
Comment (nwithan8) [reply src/main/java/com/easypost/model/SmartrateCollection.java:20]: I'm not sure I follow. Let's discuss in-person.

File: src/test/java/com/easypost/EventTest.java
Comment (nwithan8) [reply src/test/java/com/easypost/EventTest.java]: JSON works about as well as you'd expect in Java when the only connection between Java and JavaScript is [mid-90s marketing buzzwords](https://www.springboard.com/blog/data-science/history-of-javascript/#:~:text=and%2C%20later%2C%20JavaScript.-,JavaScript%20vs.%20Java,-There%E2%80%99s%20often%20some)

File: src/main/java/com/easypost/model/Fee.java
Comment (nwithan8) [reply src/main/java/com/easypost/model/Fee.java]: All boolean getters/setters docstrings like this have been clarified in aed8e7f

File: src/main/java/com/easypost/model/SmartrateCollection.java
Comment (Justintime50) [reply src/main/java/com/easypost/model/SmartrateCollection.java:20]: After chatting, let's delete this comment as it's not needed.

**Review by Justintime50 — APPROVED** [review]
Great work here! Thanks for addressing all my feedback.

**Review by jontsai — APPROVED** [review]

**Review by Justintime50 — DISMISSED** [review]

**Review by jontsai — APPROVED** [review]

**Review by Justintime50 — APPROVED** [review]

File: src/main/java/com/easypost/exception/EasyPostException.java
Comment (nwithan8) [reply src/main/java/com/easypost/exception/EasyPostException.java:41]: Will fix later, don't want to void the approvals

**Review by jchen293 — APPROVED** [review]
LGTM, nice work!

**Pull request merged**
